BLUEPRINT.md

```diff
@@ -0,0 +1,335 @@
+
+```js
+import { createVerifiedFetch } from '@helia/verified-fetch'
+
+const verifiedFetch = await createVerifiedFetch({
+  gateways: ['https://some-hash-stream-server.com'],
+})
+
+const response = await verifiedFetch(`ipfs://${cid}/`)
+const data = await response.blob()
+```
+
+---
+
+### 🛫 Lifecycle of the Response
+
+> How does the streamer response actually get assembled?
+
+#### 🕵️‍♂️ Find Index Records associated with the multihash
+
+- No need to do "content discover" — just lookup for known Index records for the multihash
+
+![image](./diagrams/hash-stream-1-index.svg)
+
+- **What’s an Index?**
+  - A tiny, server-local, map of `multihash` → `IndexRecord`.
+  - One can think of it like a `directory` or `map` for their content inside storage.
+- There are different **fundamental lenses**, which an indexing system MAY be able to index and enable queries for:
+  - A **Blob** is a set of bytes that is individually addressed by a multihash. It can be stored as is in a given store, or within a Pack.
+  - A **Pack** holds several Blobs and is also individually addressed by a multihash. Multiple **Packs** MAY include the same Blob.
+  - A **Containing** represents a set of Blobs/Packs that have some implied cryptographic relationship between them (e.g. a sharded DAG across multiple Packs/Blobs).
+
+```typescript
+type IndexRecordType = BLOB | PACK | CONTAINING
+type BLOB = 0
+type PACK = 1
+type CONTAINING = 2
+
+type IndexRecord = {
+  // MultihashDigest identifiying the record
+  multihash: MultihashDigest
+  // Type of the record
+  type: IndexRecordType
+  // hash digest of the location or Path
+  location: MultihashDigest | Path
+  // length of the data
+  length?: number
+  // offset of the data in the location byte stream
+  offset?: number
+  // associated records
+  subRecords: Array<IndexRecord>
+}
+```
+
+```js
+import { IndexReader } from '@hash-stream/index/reader'
+import { S3LikeIndexStore } from '@hash-stream/index/store/s3-like'
+
+const indexStore = new S3LikeIndexStore() // TODO: Create store with s3 client
+const indexReader = new IndexReader(indexStore)
+
+for await (const indexRecord of indexReader.findRecords(targetMultihash)) {
+  // TODO: process the index Records to fetch the bytes
+}
+```
+
+---
+
+#### 📦 Read Blobs from Pack Storage
+
+- Once the index records are found, the bytes can be read from the storage layer
+- The server may fetch the right byte ranges from a Pack (e.g., [CAR](https://ipld.io/specs/transport/car/carv1/) file)
+
+![image](./diagrams/hash-stream-2-read-blobs.svg)
+
+```js
+import { PackReader } from '@hash-stream/pack/reader'
+import { S3LikePackStore } from '@hash-stream/pack/store/s3-like'
+
+const packStore = new S3LikePackStore() // TODO: Create store with s3 client
+const packReader = new PackReader(packStore)
+
+for await (const { multihash, bytes } of this.packReader.stream(
+  targetMultihash,
+  blobRanges // Read full pack if no ranges, or just parts of the pack
```

Comment:
what do we do with `IndexRecords` to get the input for `packReader` ?

Reply:
If its just a subset of an index record I would pass the index record but to not change it now just a quick mention "get location from record and passed it to the reader" would be good

Reply:
https://github.com/vasco-santos/hash-stream/pull/103/commits/9e250192f30f0a8d9f163c07cb5b07fb4d5eabe6

Yeah, that is a good point. I will add it to a list of potential follow ups :) The reason here for clarity was that PackReader should actually work without an index system by just receiving a location, but perhaps there is some kind of common interface we can build that is compatible between the record and what PackReader needs

Reply:
Add an example of the ranges, {length, range, multihash}. 
You can clarify this multihash too because it's different from the targetMultihash

Reply:
One is hash of data bytes, this is  location hash? How do you actually hash this?

Reply:
ok probably more clear now https://github.com/vasco-santos/hash-stream/pull/103/commits/fcd25d4102704a5bc07b3b3a377fe85fdf04acea

The location hash is the hash of the PACK, so consider it like a CAR CID, and you may want to fetch a few blocks within this CAR, or just read all the blocks of it skipping the CAR internal things